src/PowerShell/PowerShellManager.cs

```diff
@@ -45,27 +46,64 @@ static PowerShellManager()
             addMethod.Invoke(null, new object[] { "HttpRequestContext", typeof(HttpRequestContext) });
         }
 
-        internal PowerShellManager(ILogger logger)
+        /// <summary>
+        /// Constructor for setting the basic fields.
+        /// </summary>
+        private PowerShellManager(ILogger logger, PowerShell pwsh, int id)
+        {
+            _logger = logger;
+            _pwsh = pwsh;
+            _pwsh.Runspace.Name = $"PowerShellManager{id}";
+        }
+
+        /// <summary>
+        /// Create a PowerShellManager instance but defer the Initialization.
+        /// </summary>
+        /// <remarks>
+        /// This constructor is only for creating the very first PowerShellManager instance.
+        /// The initialization work is deferred until all prerequisites are ready, such as
+        /// the dependent modules are downloaded and all Az functions are loaded.
+        /// </remarks>
+        internal PowerShellManager(ILogger logger, PowerShell pwsh)
+            : this(logger, pwsh, id: 1)
         {
-            if (FunctionLoader.FunctionAppRootPath == null)
+        }
+
+        /// <summary>
+        /// Create a PowerShellManager instance and initialize it.
+        /// </summary>
+        internal PowerShellManager(ILogger logger, int id)
+            : this(logger, Utils.NewPwshInstance(), id)
+        {
+            // Initialize the Runspace
+            Initialize();
+        }
+
+        /// <summary>
+        /// Extra initialization of the Runspace.
+        /// </summary>
+        internal void Initialize()
+        {
+            if (!_runspaceInited)
             {
-                throw new InvalidOperationException(PowerShellWorkerStrings.FunctionAppRootNotResolved);
+                RegisterStreamEvents();
+                InvokeProfile(FunctionLoader.FunctionAppProfilePath);
```

Comment:
This will run Profile.ps1 for every instance of PowerShellManager, is this required or does Profile.ps1 need to run once per function app?

Reply:
Yes it's required. If I define a function in the profile.ps1, I need it to be available in every PowerShellManager. The only way to do this is by running the profile.ps1 per PowerShellManager

Reply:
Yes, as @TylerLeonhardt explained, it's required.